### ISSUE TITLE
AUC Score & Precision score are different why not same?

### ISSUE THREAD

**Issue by vchouksey** [yzhao062/Pyod#2]
from pyod.utils.data import evaluate_print
# evaluate and print the results
print("\nOn Training Data:")
evaluate_print(clf_name, y_true, y_scores)

On Training Data:
KNN ROC:0.9352, precision @ rank n:0.568

from sklearn import metrics
print("Accuracy Score",round(metrics.accuracy_score(y_true, y_pred),2))
print("Precision Score",round(metrics.precision_score(y_true, y_pred),2))
print("Recall Score",round(metrics.recall_score(y_true, y_pred),2))
print("F1 Score",round(metrics.f1_score(y_true, y_pred),2))
print("Roc Auc score",round(metrics.roc_auc_score(y_true, y_pred),2))

Accuracy Score 0.92
Precision Score 0.55
Recall Score 0.59
F1 Score 0.57
Roc Auc score 0.77

**Comment by yzhao062** [comment]
I have two assumptions:
1. evaluate_print() function takes in two parameters: (i) y: binary ground truth and (ii) y_pred: **raw outlier scores** . If you pass the result from predict(), which is binary labels, would result in different results as you saw.  You may want to use **"print("Roc Auc score",round(metrics.roc_auc_score(y_true, y_scores),2))"**  instead.
2. **Precision @ rank n** is a different evaluation metric than **precision**. Sometimes it is referred as precision @ k, see this for more information.
https://medium.com/@m_n_malaeb/recall-and-precision-at-k-for-recommender-systems-618483226c54

--------------
Besides, I did a quick check as well. See the attached screenshot. Let me know if you have more questions:)
![screenshot_2](https://user-images.githubusercontent.com/15079146/40938475-f8954e62-680f-11e8-8243-d82bbe5ddb76.png)

**Comment by vchouksey** [comment]
y_train_scores is continuous values
Out[64]:
array([ 2139.81004331,   225.42848139,   239.88748023, ...,   210.58965524,
         297.59032808,   321.10746309])

y_train is binary values
y_train 
Out[65]:
array([1, 0, 0, ..., 0, 0, 0])

I have to convert either into binary or in continuous in order to compute AUC, precision score?

**Comment by yzhao062** [comment]
ROC assumes no cut-off point, so it is fine to be used with continuous values (supposed to be). Precision or Precision @ rank n is different -- they are point estimation to be used with discrete values.

Therefore, it is fine to calculate **ROC** (y=binary, y_pred=either continuous or binary) with mixed data types. In contrast, it will throw errors if you are using continuous values in calculating **precision**.

However, if you are using evaluate_print(). There is no need to do the conversion. The function handles that by calling internal utility functions to convert scores into binary labels.

![screenshot_3](https://user-images.githubusercontent.com/15079146/40940241-32d48638-6815-11e8-87e4-24a2bd7b1589.png)
![screenshot_4](https://user-images.githubusercontent.com/15079146/40940242-32e21faa-6815-11e8-8aef-6ac8730f104b.png)

**Comment by vchouksey** [comment]
Question why does precision & recall getting similar scores?

from pyod.utils.data import evaluate_print
# evaluate and print the results
print("\nOn Training Data:")
evaluate_print(clf_name, y_true, y_train_scores)
y_train_scores[y]
On Training Data:
KNN ROC:0.9352, precision @ rank n:0.568

Then I did, to binary labels for y_pred
from pyod.utils.utility import get_label_n
y_pred = get_label_n(y_true, y_train_scores)

then I did 

from sklearn import metrics
print("Accuracy Score",round(metrics.accuracy_score(y_true, y_pred),2))
print("Precision Score",metrics.precision_score(y_true, y_pred))
print("Recall Score",metrics.recall_score(y_true, y_pred))
print("F1 Score",round(metrics.f1_score(y_true, y_pred),2))
print("Roc Auc score",round(metrics.roc_auc_score(y_true, y_train_scores),2))

Accuracy Score 0.92
**Precision Score 0.568
Recall Score 0.568**
F1 Score 0.57
Roc Auc score 0.94

**Comment by vchouksey** [comment]
again on another dataset, precision & recall score is same?




Precision Score 0.836689038031
Recall Score 0.836689038031

**Comment by yzhao062** [comment]
This is expected...since you are not evaluating Precison and Recall, but Precision @ rank **n** and Recall @ rank **n**.

Once you call get_label_n(y, y_pred, n=None), y_pred (raw outlier scores) will be forced to binary outlier labels. **n specifies the number of outliers**. **If n is not set, then the number of outliers are inferred from the ground truth n.**

In other words, y_pred = get_label_n(y_true, y_train_scores) will return a binary label which contains **exactly the same number of outliers** as y_true. 

Why this matters? Simply because:

- Precision = (true positive)/(true positive+**false positive**)
- Recall = (true positive)/(true positive+**false negative**). 
The only difference between two formulas are "false positive" and "false negative". When you call get_label_n without supplying number of outliers (n), the number of false positives exactly equals to the number of false negatives. Think in this way, misclassifying one inlier as an outlier will increase both false negative and false positive by 1, since the total number of outliers are defined.

**Therefore, Precision @ rank n and Recall @ rank n are equivalent when the predicted values have exactly the same number of outliers as the ground truth.**

I make a quick demo for you. As you can see, precision and recall are only the same when we convert the predicted scores to have the same number of outliers as the ground truth. In this example, n=20 is the point, which has been highlighted.

![screenshot_5](https://user-images.githubusercontent.com/15079146/40987064-b0ce94c0-68b5-11e8-8d51-ceabdfd5df5e.png)

**Comment by vchouksey** [comment]
Thank you

**Comment by hugoas** [comment]
I just don't understand how you managed to extract F1 and Recall with the pyod lib. The concepts I already know, but I wanted to try it automatically.
Is there any module in pyod for this?
